### DISCUSSION TITLE
Stand states, emotes, emote states

### DISCUSSION THREAD

**Discussion by offl** [TrinityCore/TrinityCore#27046] [General]
Btw maybe you know how stand states are changed in scripts?
I mean is it AI action? I haven't found any action nor enum with stand states and that scares me

_Originally posted by @offl in https://github.com/TrinityCore/TrinityCore/issues/25179#issuecomment-940162712_

**Comment by offl** [comment]
_Originally posted by @xvwyh in https://github.com/TrinityCore/TrinityCore/issues/25179#issuecomment-940211319_

Interesting. My guess would be - through emotes. Emotes do have an associated stand state, so maybe if that emotestate is set on a unit - it changes the stand state instead.

<details>
<summary>example from Emote.dbc</summary>

![image](https://user-images.githubusercontent.com/43143822/136829830-6d770027-71a3-4ede-a4eb-a916d9449214.png)

</details>

That SpecProc basically defines what the emote will do: will it be a oneshot (0), a stand state change (1) or emotestate (2).

**Comment by offl** [comment]
_Originally posted by @Shauren in https://github.com/TrinityCore/TrinityCore/issues/25179#issuecomment-940213517_

SpecProc: https://wow.tools/dbc/?dbc=enumeratedstring&build=2.5.1.39640#page=1&colFilter[4]=465

**Comment by offl** [comment]
_Originally posted by @offl in https://github.com/TrinityCore/TrinityCore/issues/25179#issuecomment-940369255_

Indeed, thanks
Yeah, that's super scary

>Emote State - Set and Forget (formerly perform an emote)
>Emote State with Idle (formerly continually emote)

Pehaps that's why they has the 'formerly' comment, they're really old and were merged long time ago
And all emotes has prefixes to easily select the one you need in WoWEdit (ONESHOT, STATE, one EMOTE_STAND_STATE_NONE)

Looks like we need support for SpecProc
I wonder what value they uses to clean both stand and emote states, 0 or EMOTE_STATE_NONE (weird)

**Comment by offl** [comment]
_Originally posted by @xvwyh in https://github.com/TrinityCore/TrinityCore/issues/25179#issuecomment-940445878_

There's ONESHOT_NONE, STATE_NONE and STAND_STATE_NONE. Each for its own task, I suppose.

>  both stand and emote states

I doubt "both" are ever set at the same time. If you attempt to play an emote with "One Shot Emote" SpecProc - it will just send the opcode. If you attempt to play an emote with "Set Emote State" SpecProc - it'll set the emotestate but won't send the oneshot emote opcode. If you attempt to play an emote with "Changes Stand State" - it will change stand state but won't affect the emotestate or send the emote opcode.

I also wondered in the past why is there STATE_NONE which isn't 0. Guess this is why.

**Comment by offl** [comment]
_Originally posted by @offl in https://github.com/TrinityCore/TrinityCore/issues/25179#issuecomment-940467658_

SpecProc looks good for NONE https://wow.tools/dbc/?dbc=emotes&build=3.3.5.12340#page=1&colFilter[1]=none
But EMOTE_STAND_STATE_NONE is at very bottom of enum

Also https://github.com/TrinityCore/TrinityCore/commit/1c174987c8114d7f2daf02140afa81a1cb01bd54#diff-dbc570c8e18c401307e01fa0c5ecb5ed459fa364c2e16b1e66e37459a24d750eR14-R16
And https://github.com/TrinityCore/TrinityCore/commit/c737817a600dbb33b070f4172052f9ace81d8097#diff-a8187238022b7502a10c810064112dc5bde2fcc06a8629d0bea6f6c7f0442b60R7-R8
Are the main reason why I started to think what we can do with all that mess(action to change stand states for some reason is SetByteFlag and not setvalue > you need to remove current stand state before adding new or client not gonna like it)

4 states are not stand states (no 4,5,6,9) https://wow.tools/dbc/?dbc=emotes&build=3.3.5.12340#page=1&colFilter[1]=sit_chair
https://wow.tools/dbc/?dbc=emotes&build=3.3.5.12340#page=1&colFilter[1]=submerge
But I guess we can figure out what to do

**Comment by offl** [comment]
_Originally posted by @xvwyh in https://github.com/TrinityCore/TrinityCore/issues/25179#issuecomment-940498542_

> But EMOTE_STAND_STATE_NONE is at very bottom of enum

Because normally you can just use STATE_STAND to achieve the same effect. Maybe they needed STAND_STATE_NONE for some technical reason. Who knows.

Another possible reason: STATE_STAND wasn't working for them because of its flags, so they made a STAND_STATE_NONE with different flags. No idea why, but hey, that's the only part which they differ in.

> action to change stand states for some reason is SetByteFlag and not setvalue

Because "action to change stand states" doesn't exist. 90 and 91 are general purpose SMART_ACTION_SET_UNIT_FIELD_BYTES_1/SMART_ACTION_REMOVE_UNIT_FIELD_BYTES_1. UNIT_FIELD_BYTES_1 houses two other bytes, one of which definitely is flags, and the other one was presumed to be flags until a few years ago, so it was "reasonable" to make the action deal with flags instead. But yeah, ideally they should be reworked into "Set Stand State" and "Set Animation Tier" (assuming there isn't one already) actions, and `enum UnitStandFlags` - or whatever it is at bytes1[2] - should not be touched from SAI to begin with.

> 4,5,6

Chair poses for different height of chairs. I'd say it's highly likely they simply aren't used in scripts.

> 9

Every time I see "submerged" I flinch. It's a bloody mess. There's stand state for "submerged" - it doesn't work half of the time. There's emote state for "submerged" - it doesn't work half of the time, but it's a *different* half of the time to the first one. There's animation tier "Submerged" with a comment "(NOT YET IMPLEMENTED)", and in some later expansion it gets removed:
https://wow.tools/dbc/?dbc=enumeratedstring&build=2.5.1.39640#page=1&colFilter[4]=705 - it's called "Obsolete2" now. If you make a creature play the "Submerg**e**" animation via spell, and change the standstate/emotestate to "Submerge**d**" at a wrong time - the creature will either briefly reappear or the animation will cancel itself. I guess blizzard were really trying to get it to work (likely for northrend jormungar wurms and the like) but it just wasn't working out for them.

Now, why are there STAND_STATE_SUBMERGED, STATE_MOUNT and STATE_READYJOUST, which have SpecProc="Changes Stand State" BUT ALSO have an animation - THAT is a good question...

**Comment by xvwyh** [comment]
I can explain the difference between these two actions that you mentiond:

> Emote State - Set and Forget (formerly perform an emote)
> Emote State with Idle (formerly continually emote)

Blizzard's scripting system works like coroutines: they have a plain list of actions that are executed in order, after the previous action finishes. For example:
1. Walk to point A
2. Speak text
3. Wait for 5 seconds
4. Walk to point B
5. Cast a spell
6. Walk to point C
7. Wait for 10 seconds
8. Loop

That "Wait for X seconds" is called "[Idle](https://docs.google.com/spreadsheets/d/1Wh1e1ZXYM-sw3wWrY9dGZ2ngK8oa3PA2XIftQJuf5R0/edit#gid=740715654&range=A9:B10)". Idle for X seconds.

The difference between the two aforementioned actions would be:
- **Emote State - Set and Forget (formerly perform an emote)** - executes the emote (oneshot or sets the emotestate or stand state if my assumptions are correct) and immediately moves on to the next step of the script. Emotestate/standstate stays active until manually set to none with this same action type.
- **Emote State with Idle (formerly continually emote)** - executes the emote, waits the specified amount of time, stops the emote, proceeds to the next script step.

Why did they rename it from a generic "emote" to "emote state" is beyond me. This is the only action capable of dealing with emotes, so it's weird if they decided to make it stop being usable for oneshot emotes. But I have a suspicion that since they are now able to tie BroadcastText with corresponding emotes (and sounds, both of which they couldn't do in Classic) - that action stopped being relevant for oneshot emotes, since they can now bundle them directly with the text and they don't need to manually script creatures to "1. speak text 2. play sound 3. play emote" with 3 separate actions.

**Comment by offl** [comment]
>>action to change stand states for some reason is SetByteFlag and not setvalue
>
>Because "action to change stand states" doesn't exist. 90 and 91 are general purpose SMART_ACTION_SET_UNIT_FIELD_BYTES_1/SMART_ACTION_REMOVE_UNIT_FIELD_BYTES_1. UNIT_FIELD_BYTES_1 houses two other bytes, one of which definitely is flags, and the other one was presumed to be flags until a few years ago, so it was "reasonable" to make the action deal with flags instead. But yeah, ideally they should be reworked into "Set Stand State" and "Set Animation Tier" (assuming there isn't one already) actions, and enum UnitStandFlags - or whatever it is at bytes1[2] - should not be touched from SAI to begin with.

Indeed... Yeah, I wanted to add the Set Animation Tier

>Chair poses for different height of chairs. I'd say it's highly likely they simply aren't used in scripts.

Sit chair stand states are shown in sniffs, they're used atm in scripts but I don't know if they're sniffed or not
If you mean they're not removed or set in scripts, they need a way to add them as 'default emote' (if such thing exists)
So they need to use it as emote and not as stand state, aren't they?

>>I wonder what value they uses to clean both stand and emote states, 0 or EMOTE_STATE_NONE (weird)
>
>There's ONESHOT_NONE, STATE_NONE and STAND_STATE_NONE. Each for its own task, I suppose.

Hol up, why they need to remove ONESHOT?
This is probably the real problem here, it can be hard to remove stand and emote states using not 0 but specific emote

**Reply by xvwyh** [reply]
> Sit chair stand states are shown in sniffs, they're used atm in scripts but I don't know if they're sniffed or not

Are you sure it's the UNIT_STAND_STATE_SIT_LOW_CHAIR, UNIT_STAND_STATE_SIT_MEDIUM_CHAIR or UNIT_STAND_STATE_SIT_HIGH_CHAIR, and not just a regular UNIT_STAND_STATE_SIT_CHAIR?

> So they need to use it as emote and not as stand state, aren't they?

I guess? If my theory is correct then yes.

> Hol up, why they need to remove ONESHOT?

Another great question!
...Moving on...

> This is probably the real problem here, it can be hard to remove stand and emote states using not 0 but specific emote

No, it can't be hard. One way or another, if you want to start using SpecProc - you'll need to introduce a function that checks the dbc data for the emote passed to it and chooses how to proceed based on it. Ain't hard to put a special clause for "Sets Emote State" case, e.g.:

```cpp
void Unit::PerformEmote(Emote emote)
{
    EmotesEntry const* entry = sEmotesStore.LookupEntry(emote);
    if (!entry)
        return;

    switch (entry->EmoteType)
    {
        case 0:
            HandleEmoteCommand(emote);
            break;
        case 1:
            SetStandState(entry->UnitStandState);
            break;
        case 2:
            SetUInt32Value(UNIT_NPC_EMOTESTATE, emote == EMOTE_STATE_NONE ? 0 : emote);
            break;
    }
}
```

The biggest problem would honestly be getting your scripters to use the proper emote for their task. Far too many would just keep using 0 to stop the emotestate and will start complaining when it stops working. Explaining how to pick the correct emote will be difficult without proper tools. And even after explaining they might start complaining about why did we overcomplicate such a simple system.

**Reply by offl** [reply]
>>Sit chair stand states are shown in sniffs, they're used atm in scripts but I don't know if they're sniffed or not
>
>Are you sure it's the UNIT_STAND_STATE_SIT_LOW_CHAIR, UNIT_STAND_STATE_SIT_MEDIUM_CHAIR or UNIT_STAND_STATE_SIT_HIGH_CHAIR, and not just a regular UNIT_STAND_STATE_SIT_CHAIR?

Yep, checked again sniffs, 4, 5 and 6 are shown in sniffs
I think this is sniffed https://tcubuntu.northeurope.cloudapp.azure.com/aowow/?npc=20345#sai-actionlist-2034500
But I don't have sniff with that event and have no idea what action 38 does and how it looked in sniffs
Action 61 changes stand state to 4. But again, I don't have even sniff with spawn of that creature to confirm it's indeed 4 as default

>The biggest problem would honestly be getting your scripters to use the proper emote for their task. Far too many would just keep using 0 to stop the emotestate and will start complaining when it stops working. Explaining how to pick the correct emote will be difficult without proper tools. And even after explaining they might start complaining about why did we overcomplicate such a simple system.

Yes, that is what I meant here > This is probably the real problem here, it can be hard to remove stand and emote states using not 0 but specific emote
I guess it's not that bad if 0 is 'remove everything' and we figure out what to do with chair and submerge stand states

**Reply by Shauren** [reply]
Your "special clause" doesn't work, because it goes to case 0 when you give it EMOTE_STATE_NONE

**Reply by xvwyh** [reply]
> Yep, checked again sniffs, 4, 5 and 6 are shown in sniffs

Then I have no clue.

> Your "special clause" doesn't work, because it goes to case 0 when you give it EMOTE_STATE_NONE

Incorrect.
```
    EMOTE_ONESHOT_NONE                 = 0,   (SpecProc=0)
    EMOTE_STATE_NONE                   = 30,  (SpecProc=2)
    EMOTE_STAND_STATE_NONE             = 470, (SpecProc=1)
```

I did make a mistake of having case 1 and 2 flipped (correcting now), but case 0 will not be hit with EMOTE_STATE_NONE.

**Reply by offl** [reply]
SpecProc was implemented somewhere in alpha or vanilla(doesn't really matter)
But
>EMOTE_STAND_STATE_NONE             = 470, (SpecProc=1)

Didn't existed in 3.0.1

The one which looks correct is STATE_STAND (26)
With STATE_NONE (30)

But again ONESHOT_NONE looks weird in that list
Maybe ONESHOT_NONE is just like default value in enums?

***

If we make a mistake and will use 0 as 'remove all', we'll be not able to easily start using 26 and 30 because we'll not know what we need to remove - stand state or emote state
Looks like the way those actions are handled in SAI makes it easy to merge them in one

**Reply by xvwyh** [reply]
Another possibility how these special chair standstates can be achieved, albeit a far fatched one: maybe they just interact with chairs like players do? There is an "Unit(s) activate a game object" action. And chairs have a property that specifies which standstate should be used with them.

But if this is the case, then some other conditions need to happen:
- There must be a chair-type gameobject near a unit that has such a standstate.
- The moment they change their standstate (or rather a little bit earlier - updatefields are delayed) there should also be an accompanying teleport/movementupdate opcode that moves them to the respective chair seat, otherwise you would need to argue that chairs have different implementation for creatures and players and this would make this theory very weak.

**Reply by offl** [reply]
I was thinking about something like that
Eh, if that is true it's something really bad for us. We can simply hack it and leave for better days or we may never start use anything of that
This way we'll use stand states as emotes and if I understand correctly this way we'll start use emote flags in case of stand states, while currently we don't use them

May-be-good question: can player sit on the same chair NPC sit on?

To me using STATE_STAND and STATE_NONE doesn't look as bad as before

**Reply by xvwyh** [reply]
> May-be-good question: can player sit on the same chair NPC sit on?

Up to the server to decide. The client will allow you to interact with the chair no matter what, except when the local player itself is in one of the seat's coordinates and has one of those special standstates - in this case the chair is not clickable, to make it look like you're in the process of interacting with it.

The current TC implementation is not a 100% match to the client's code, but it's good enough.
https://github.com/TrinityCore/TrinityCore/blob/07d1b5e64b7de969ce2e287d540f3faa501381de/src/server/game/Entities/GameObject/GameObject.cpp#L1532
It doesn't support non-player users right now, but adding support for them is trivial, if need be.
